kotlin-inject-compiler-core/src/main/kotlin/me/tatarka/inject/compiler/CycleDetector.kt

```diff
@@ -2,25 +2,57 @@ package me.tatarka.inject.compiler
 
 class CycleDetector {
 
-    private val _elements = mutableListOf<AstElement>()
+    private val _elements = mutableListOf<Entry>()
+    private var resolving: CycleResult.Resolvable? = null
 
-    val elements: List<AstElement> get() = _elements
+    val elements: List<AstElement>
+        get() = _elements.mapNotNull {
+            when (it) {
+                is Entry.Element -> it.value
+                else -> null
+            }
+        }
+
+    fun delayed() {
+        _elements.add(Entry.Delayed)
+    }
 
-    fun check(element: AstElement): CycleResult {
-        return if (element in _elements) {
-            // TODO: check resolvable
-            CycleResult.Cycle
+    fun resolve(key: TypeKey): Boolean {
+        return if (resolving?.key == key) {
+            resolving = null
+            true
         } else {
-            _elements.add(element)
+            false
+        }
+    }
+
+    fun check(key: TypeKey, element: AstElement): CycleResult {
+        return if (_elements.any { it is Entry.Element && it.value == element }) {
+            if (_elements.any { it is Entry.Delayed }) {
+                CycleResult.Resolvable(key).also { resolving = it }
+            } else {
+                CycleResult.Cycle
+            }
+        } else {
+            _elements.add(Entry.Element(element))
             CycleResult.None
         }
     }
 
     fun pop() {
```

Comment:
Not sure if this is a good name for it anymore. Since it may pop more than one thing

Reply:
Cleaned up so the pop method is no longer needed.